pep-0497-externals.txt

```diff
@@ -0,0 +1,460 @@
+inference layer being implemented or specified. Alternatively, a separate PEP
+could be developed in tandem with this proposal to hold the apppropriate level
+of detail.
+
+
+Types of environment and vendor
+-------------------------------
+
+Anaconda is not an operating system. Neither is yum, neither is apt. It is the
+maintainer of the packaging system which is pertinent, and the current operating
+system is an imperfect indicator of the relevant package manager.
+
+
+Directories, categories and locations of dependency files
+---------------------------------------------------------
+
+The proposed mechanism allows introspection of the standard location for
+packages. On *nix environments, these would be (for example), /usr/bin,
+/usr/local/include etc (see https://www.gnu.org/prep/standards/html_node/Directory-Variables.html
+for a discussion of standard directories). Non-standard dependency locations
+would be supported only via the underlying platform's mechanisms for exposing
+these details, such as PATH, LD_LIBRARY_PATH etc. This PEP does not include
+generalised configuration and specification of dependencies in non-standard
+or customised locations.
+
+The Python installation tools should honour those standard mechanisms.
+The details are per-platform and not specified by this specification
+(see Platform Abstraction for more on this point).
+
+Virtual Environment Concerns
+----------------------------
+
+This should work fine with virtual environments. Since the PEP draws the line
+at performing external package installation, it is up to package management
+tools to decide how to separate concerns between systemwide installation
+and virtual environment installation. For example, a system such as
+Anaconda which is heavily based around virtual environments is well capable
+of handling the installation of external dependencies into a virtual environment
+context.
+
+Alternative Syntax Concepts
+==========================
+
+Concept 0: Separated external and non-external dependencies
+-----------------------------------------------------------
+
+This is an early-stage syntax. Concept 1 replaces it as the current
+"front runner" as implementation candidate and is included in the abstract.
+
+The advantage of Concept 0 is that is clearly separates the non-Python
+dependency specification into an isolated part of the configuration arguments,
+reducing the potential for confusion between external and internal dependencies.
+This may assist during implementation as no 'sniffer' will be needed to
+determine whether a named package is internal or external.
+
+::
+
+    setup(
+        name='uses_scipy_with_atlas_extension'
+        install_requires='yaml, scipy[atlas]',
+        setup_requires='pyfortran',
+        external={
+          setup=[
+            'bin!fortran:dist_name=='ubuntu',
+            'include!kernel.h',
+            'include!liblas.h',
+          ],
+          install=[
+              'lib!liblas1.so'
```

Comment:
Hi Leonardo,

I am coming to this discussion a bit late, because "life". I was involved
in authoring this PEP although rbt now has carriage of it.

We discussed the issues you mentioned in some depth. I am not myself a
packaging guru or compile expert, but I am a maintainer of small packages
and software engineer. I'm a 'user' of pip for the most part. You're not
wrong in identifying those concerns but I think there is a balance of
issues to consider. I will present my viewpoint based on my use cases:

  (i) Firstly, my view is that the ability to store these things in
setup.py will more or less immediately address a very large proportion of
the really common issues and is well worth doing.

  (ii) Engineering the relevant build tools is much harder than allowing
some simple manual configuration

  (iii) Pulling the dependency gaps out of build tools might work for the
maintainer, but you would still need to store the dependency gap
information in some kind of file. In my use case, the install will often
fail 90 seconds into compiling some dependency, and the goal is to provide
the information back to users much faster than that. This isn't about
'fixing dependency resolution in across wheels and external packaging
systems (e.g. yum/conda), but rather about providing the pragmatic
workarounds required to fix the major pain points.

  (iv) A previous accepted PEP did include some capability to include
metadata in one of the metadata files. To my knowledge it was never
implemented in tools. I think it was just too complicated. On the other
hand, I know of a lot of immediate issues a simple extension to setup.py
would allow

In general, the uses I know of are for 'the exception' rather than 'the
rule'. For example, packages like scipy are "mostly fine" to compile, but
fail in some environments.

I would also consider this might not be mutually exclusive with what you
are suggesting. It may be that the larger engineering work you are
suggesting is justified, I am not sure. I would certainly love it if there
were a two-way bridge between pip and the various external packaging
systems so that e.g. a pip install could also submit its dependencies to
yum for a request to install, however we just don't live in that world.

Thanks,
-Tennessee

On 14 August 2015 at 01:44, Leonardo Rochael Almeida <
notifications@github.com> wrote:

> In pep-0497-externals.txt
> <https://github.com/pypa/interoperability-peps/pull/30#discussion_r36989624>
> :
>
> > +determine whether a named package is internal or external.
> > +
> > +::
> > +
> > +    setup(
> > +        name='uses_scipy_with_atlas_extension'
> > +        install_requires='yaml, scipy[atlas]',
> > +        setup_requires='pyfortran',
> > +        external={
> > +          setup=[
> > +            'bin!fortran:dist_name=='ubuntu',
> > +            'include!kernel.h',
> > +            'include!liblas.h',
> > +          ],
> > +          install=[
> > +              'lib!liblas1.so'
>
> Specifying install dependencies here on setup.py will be very useful,
> specially with regards to specifying external processes that will be needed
> by the package (e.g. bin!pdftk).
>
> However, regarding dependencies into external shared libraries, it would
> be better if the build tool generated those automatically.
>
> It's a lot easier for, say, setuptools or an extension to it to run ldd
> build/lxml_***/etree.so and figure out the library dependencies and put
> those in the wheel metadata, than forcing the poor package maintainer to do
> so to put it into setup.py.
>
> In particular, different linux distros will have libraries that are source
> compatible at build time, but result in shared library dependencies with
> different versions on the python extension.
>
> Forcing this information to be present in setup.py would force the
> package maintainer to build this package in a variety of platforms just to
> be able to add the proper 'lib!libsomenthing.so.X.Y:dist_name=="Z"' for
> every supported platform.
>
> Whereas if the tooling automatically fed this information to the wheel
> metadata, not only would the package maintainer be relieved of this chore,
> but the package builder would be able to chose between different source
> compatible libraries, or even between static and dynamic linking, and the
> generated wheel would correctly carry the information needed to install it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pypa/interoperability-peps/pull/30/files#r36989624>.
>



-- 
--------------------------------------------------
Tennessee Leeuwenburg
http://myownhat.blogspot.com/
"Don't believe everything you think"